File: DFS&BFS/python/1963.py
from collections import deque
t = int(input())
visited = [False for _ in range(10000)]

for i in range(2,int(9999**0.5)+1):
  j = 2
  while (i*j <=9999):
    visited[i*j] = True
    j+=1

sosoo = set(i for i in range(1000,10000) if visited[i] == False)

for _ in range(t):
  num,target = map(int,input().split())
  q = deque()
  q.append((num,0))
  result = "Impossible"
  visited = set()
  visited.add(num)
  while q:
    num,cnt = q.popleft()
    if num == target:
      result = cnt
      break
    for i in range(10):
      arr = []
      arr.append((num//10)*10+i)
      arr.append((num//100)*100 + (num%10) + (i*10))
      arr.append((num//1000)*1000 + (num%100) + i*100)
      if i != 0: arr.append((num%1000)+ (i*1000))
      for next_num in arr:
        if next_num not in sosoo: continue
        if next_num in visited: continue
        visited.add(next_num)
        q.append((next_num,cnt+1))
  print(result)

File: DFS&BFS/javascript/1963.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const visited = new Array(10000);

for (let i = 2; i <= Math.floor(9999 ** 0.5) + 1; i++) {
  let j = 2;
  while (i * j <= 9999) {
    visited[i * j++] = true;
  }
}

sosoo = new Set();
for (let i = 1000; i <= 9999; i++) {
  if (!visited[i]) sosoo.add(i);
}

for (let i = 1; i < input.length - 1; i++) {
  let idx = 0;
  let [a, b] = input[i].split(" ").map(Number);
  if (a === b) {
    console.log(0);
    continue;
  }
  const q = [];
  let result = "Impossible";
  const visited = new Set();
  visited.add(a);
  q.push([a, 0]);
  while (q.length > idx) {
    let [num, cnt] = q[idx++];
    for (let i = 0; i < 10; i++) {
      const arr = [];
      arr.push(~~(num / 10) * 10 + i);
      arr.push(~~(num / 100) * 100 + i * 10 + (num % 10));
      arr.push(~~(num / 1000) * 1000 + i * 100 + (num % 100));
      arr.push((num % 1000) + i * 1000);
      for (let i = 0; i < arr.length; i++) {
        if (arr[i] === b) {
          result = cnt + 1;
          break;
        }
        if (!sosoo.has(arr[i])) continue;
        if (visited.has(arr[i])) continue;
        visited.add(arr[i]);
        q.push([arr[i], cnt + 1]);
      }
      if (result !== "Impossible") break;
    }
    if (result !== "Impossible") break;
  }
  console.log(result);
}
